snap/snapcraft.yaml

```diff
@@ -24,43 +25,125 @@ layout:
+      # interpreter depends on the build environment. Notably, local builds
+      # (`/root`) differ from builds on launchpad (`/build`).
+      sed --in-place \
+        --expression "1 s:^#\!/.*$:#\!/snap/hotsos/current/usr/bin/${PYTHON}:" \
+        ${SNAPCRAFT_PART_INSTALL}/bin/hotsos
+
+      # Add a missing link in snap to Python interpreter.
+      ln -sf ../usr/bin/${PYTHON} ${SNAPCRAFT_PART_INSTALL}/bin/python3
+
+      # Do not use `craftctl default` and `organize` because this simply copies
+      # the whole repository to the install directory. Using rsync on `defs` is
+      # cleaner.
+      mkdir -p ${SNAPCRAFT_PART_INSTALL}/etc/hotsos
+      rsync -av defs ${SNAPCRAFT_PART_INSTALL}/etc/hotsos/
+    stage:
       # need to exclude these files since they are dangling symlinks
       # that cause the snapstore to reject the upload.
-      exclude-systemd-files:
-        - -lib/systemd/system/hwclock.service
-        - -lib/systemd/system/rc.service
-        - -lib/systemd/system/x11-common.service
-        - -lib/systemd/system/cryptdisks-early.service
-        - -lib/systemd/system/rcS.service
-        - -lib/systemd/system/cryptdisks.service
-    stage:
-      - $exclude-systemd-files
+      - -lib/systemd/system/cryptdisks-early.service
+      - -lib/systemd/system/cryptdisks.service
+      - -lib/systemd/system/hwclock.service
+      - -lib/systemd/system/rc.service
+      - -lib/systemd/system/rcS.service
+      - -lib/systemd/system/x11-common.service
+      - -tests
+
+  patchelf:
+    after:
+      - hotsos
+    plugin: nil
+    build-packages:
+      - file
+      - patchelf
+    override-prime: |
+      # patchelf
+      #
+      # To allow for a classic snap that works across multiple operating system
+      # runtime environments, we need to ensure all shipped binaries consume
+      # libraries from the core or shipped snap.  We accomplish this by patching
+      # `RPATH`` or interpreter into dynamically linked binaries.
+      #
+      # /snap/${CORE}/current/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+      # /snap/${CORE}/current/lib/aarch64-linux-gnu/ld-linux-aarch64.so.1
+      # /snap/${CORE}/current/lib/powerpc64le-linux-gnu/ld64.so.2
+      # /snap/${CORE}/current/lib/s390x-linux-gnu/ld64.so.1
+      CORE=core20
+      interpreter_prefix=/snap/${CORE}/current/lib/${SNAPCRAFT_ARCH_TRIPLET}
+      snap_prefix=/snap/${SNAPCRAFT_PROJECT_NAME}/current
+      ld_arch=$(echo ${SNAPCRAFT_TARGET_ARCH} | sed -e 's,amd64,x86-64,' -e 's,arm64,aarch64,')
+
+      for interpreter in "ld-linux-${ld_arch}.so.?" "ld64.so.?" "/dev/null"; do
+          if [ -e ${interpreter_prefix}/${interpreter} ]; then
+              break
+          fi
+      done
+      if [[ ${interpreter} == /dev/null ]]; then
+          echo "Unable to determine name of shared library interpreter."
+          exit 1
+      fi
+
+      for binary in \
+        $(find . -exec file {} \; | awk -F : '/ELF.*dynamic/ { print $1 }'); do
+          patchelf \
```

Comment:
I think this loop could use some log output (e.g. setting rpath of file %s to %s).

Reply:
I added `set -x` to the script. Does that address your logging comment?